### PR TITLE
[full ci] Certificate generation and TLS verify

### LINKED ISSUE #2197: Add user authentication to the personality
The default docker mechanism is TLSVERIFY. This would entail:
a. modification to vic-machine to supply CA certificate to validate clients

There is also the extended docker authorization plugin mechanism which presumably requires that the client send user credentials (or session identifier) with each request. I do not know the details of this mechanism and a quick search of the remote API 1.24 docs didn't make it clear.

Tagged as vic-machine because any approach will require vic-machine changes for configuration. Estimate is for TLSVERIFY - see vicadmin for current mechanism.

https://bugzilla.eng.vmware.com/show_bug.cgi?id=1727625

### PR DESCRIPTION
**Pull request by hickeng**
Fixes #2197 

This expands the certificate generation in vic-machine to include creation
or use of a CA, and creation or use of signed certificates.
This enables user authentication via client certificates

This does require either a known IP address or a known DNS name (can use
wildcards for a single FQDN element, e.g. *.vmware.com, but not
*.*.vmware.com)

Still pending are all of the updates to Robot to make use of this - blocked
by either predictable DNS names or static IPs

### CODE REVIEW

**Comment by hickeng** [comment]
Known issue:
a. vic-machine uses the IP address to connect after deploy to confirm components have started.
b. you **must** connect using FQDN that matches the cname in the certificate, or
c. add the IP address of the host to the cert

This results in the component validation failing, but successful connection when using the FQDN.

**Comment by hickeng** [comment]
Having to update vic-machine to generate the certificates into a subdir along with an env file to use them - this is so that we have a structured way of setting up the env for tests and --tlsverify. End result looks as follows:
```
INFO[2016-10-06T02:45:20Z] Generating CA certificate/key pair - private key in ./envtest/ca-key.pem
INFO[2016-10-06T02:45:21Z] Generating server certificate/key pair - private key in ./envtest/server-key.pem
INFO[2016-10-06T02:45:21Z] Generating client certificate/key pair - private key in ./envtest/key.pem
INFO[2016-10-06T02:45:21Z] ### Installing VCH ####
INFO[2016-10-06T02:45:23Z] Validating supplied configuration
INFO[2016-10-06T02:45:26Z] Firewall status: DISABLED on "/ha-datacenter/host/office2-sfo2-dhcp327.eng.vmware.com/office2-sfo2-dhcp327.eng.vmware.com"
INFO[2016-10-06T02:45:26Z] Firewall configuration OK on hosts:
INFO[2016-10-06T02:45:26Z]   "/ha-datacenter/host/office2-sfo2-dhcp327.eng.vmware.com/office2-sfo2-dhcp327.eng.vmware.com"
INFO[2016-10-06T02:45:26Z] License check OK
INFO[2016-10-06T02:45:26Z] DRS check SKIPPED - target is standalone host
INFO[2016-10-06T02:45:29Z] Creating Resource Pool "envtest"
INFO[2016-10-06T02:45:29Z] Creating VirtualSwitch
INFO[2016-10-06T02:45:29Z] Creating Portgroup
INFO[2016-10-06T02:45:30Z] Creating appliance on target
INFO[2016-10-06T02:45:30Z] Network role "client" is sharing NIC with "external"
INFO[2016-10-06T02:45:30Z] Network role "management" is sharing NIC with "external"
INFO[2016-10-06T02:45:32Z] Uploading images for container
INFO[2016-10-06T02:45:32Z]      "bootstrap.iso"
INFO[2016-10-06T02:45:32Z]      "appliance.iso"
INFO[2016-10-06T02:45:38Z] Waiting for IP information
INFO[2016-10-06T02:45:45Z] Waiting for major appliance components to launch
INFO[2016-10-06T02:45:48Z] Initialization of appliance successful
INFO[2016-10-06T02:45:48Z]
INFO[2016-10-06T02:45:48Z] vic-admin portal:
INFO[2016-10-06T02:45:48Z] https://10.118.69.36:2378
INFO[2016-10-06T02:45:48Z]
INFO[2016-10-06T02:45:48Z] Docker environment variables:
DOCKER_TLS_VERIFY=1
DOCKER_CERT_PATH=/home/vagrant/vic/bin/envtest
DOCKER_HOST=office2-sfo2-dhcp142.eng.vmware.com:2376
INFO[2016-10-06T02:45:48Z] Environment saved in envtest/envtest.env
INFO[2016-10-06T02:45:48Z]
INFO[2016-10-06T02:45:48Z] Connect to docker:
INFO[2016-10-06T02:45:48Z] docker -H office2-sfo2-dhcp142.eng.vmware.com:2376 --tlsverify --tlscacert=./envtest/ca.pem --tlscert='./envtest/cert.pem' --tlskey='./envtest/key.pem' info
INFO[2016-10-06T02:45:48Z] Installer completed successfully
```

The _DOCKER_XXX_ environment variables in the output have been dropped onto lines without the timestamp so that they are easy to copy/paste. This would be even simpler if we were using a separate logger for outputting to the user as we could prefix the commentary with shell comment characters to allow piping & carelessness

File: cmd/vic-machine/create/create.go [hunk 1]
```diff
@@ -326,21 +371,6 @@ func (c *Create) Flags() []cli.Flag {
 	return flags
 }
 
-func (c *Create) processVolumeStores() error {
```
Comment (hickeng) [inline cmd/vic-machine/create/create.go]: moved down below the main processParams to live with the other processX methods.

**Comment by hickeng** [comment]
Compose test is failing currently

**Comment by andrewtchin** [comment]
lgtm

File: cmd/vic-machine/create/create.go [hunk 2]
```diff
@@ -547,28 +673,135 @@ func (c *Create) processInsecureRegistries() error {
...
+	return certs, keypair, nil
+}
+
+func (c *Create) generateCertificates(ca bool) ([]byte, *certificate.KeyPair, error) {
+	defer trace.End(trace.Begin(""))
+
+	var certs []byte
+	// generate the certs and keys with names conforming the default the docker client expects
+	// to avoid overwriting for a different vch, place this in a directory named for the vch
+	err := os.MkdirAll(fmt.Sprintf("./%s", c.DisplayName), 640)
```
Comment (dougm) [inline cmd/vic-machine/create/create.go]: Should we use 600 for the perms, or do you want group read access?
And `Mkdir` instead of `MkdirAll`, so we don't overwrite if already exists?  Maybe use `MkdirAll` only if `-force` flag is given.

File: cmd/docker/main.go [hunk 3]
```diff
@@ -218,6 +218,7 @@ func startServerWithOptions(cli *CliOptions) *apiserver.Server {
 			// server requires and verifies client's certificate
 			tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
 			tlsConfig.ClientCAs = loadCAPool()
+			tlsConfig.InsecureSkipVerify = true
```
Comment (dougm) [inline cmd/docker/main.go]: Shouldn't this be `false` here?  If not, can you add a comment where/when it'll get set to false?

File: lib/install/management/inspect.go [hunk 4]
```diff
@@ -71,16 +74,43 @@ func (d *Dispatcher) ShowVCH(conf *config.VirtualContainerHostConfigSpec, key st
...
 	if !conf.HostCertificate.IsNil() {
 		// if we're generating then there's no CA currently
-		if len(conf.CertificateAuthorities) > 0 && key != "" {
-			tls = fmt.Sprintf(" --tls --tlscert='%s' --tlskey='%s'", cert, key)
+		if len(conf.CertificateAuthorities) > 0 {
+			// find the name to use
+			addr, _ = addrToUse(d.HostIP, conf)
+
+			if key != "" {
+				tls = fmt.Sprintf(" --tlsverify --tlscacert=%s --tlscert='%s' --tlskey='%s'", cacert, cert, key)
```
Comment (dougm) [inline lib/install/management/inspect.go]: Should quote tlscacert option too.  Can also just use `%q` instead of `'%s'`.

File: pkg/certificate/certificate.go [hunk 5]
```diff
@@ -116,40 +194,110 @@ func (k *Keypair) generate() error {
 		return err
 	}
 
-	keyFile, err := os.Create(k.keyFile)
+	keyFile, err := os.OpenFile(kf, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
```
Comment (dougm) [inline pkg/certificate/certificate.go]: Suppose you could ignore my dir permissions comment above, since the key itself is not readable by the group.

File: tests/local-integration-test.sh [hunk 6]
```diff
@@ -35,6 +35,16 @@ if [ -z "$GITHUB_TOKEN" ] || [ -z "$GOVC_URL" ]; then
     exit 1
 fi
 
+# check if govc env command works as expected
+if [[ $(govc env GOVC_URL_SCHEME) =~ GOVC ]]; then
```
Comment (dougm) [inline tests/local-integration-test.sh]: Could just check the latest release version is being used: govc version -require 0.9.0

Comment (hickeng) [inline tests/local-integration-test.sh]: I had no idea this flag existed!

File: cmd/vic-machine/create/create.go (hunk 2 above)
Comment (hickeng) [inline cmd/vic-machine/create/create.go]: @dougm MkdirAll doesn't overwrite anything - can you rephrase?

**Comment by fdawg4l** [comment]
lgtm.

Stared at this change for a long time and realized i know very little about security.  Looks fine to me, though :-).

File: cmd/vic-machine/create/create.go (hunk 2 above)
Comment (dougm) [inline cmd/vic-machine/create/create.go]: `MkdirAll` itself doesn't overwrite, but doesn't error if the directory already exists - if it does already exist the file creates that follow would overwrite.  Whereas `Mkdir` fails if the directory already exists.

**Comment by hickeng** [comment]
poke

**Comment by hickeng** [comment]
poke
